CMakeLists.txt

```diff
@@ -0,0 +1,162 @@
+      else()
+          message(FATAL_ERROR "MPI f90 module not found and MPI f08 module not found")
+      endif()
+  elseif(with_mpih)
+      # use f77 interface if explicitly requested
+      if(MPI_Fortran_HAVE_F77_HEADER)
+          add_compile_definitions("MPIH")
+      else()
+          message(FATAL_ERROR "Fortran 77 mpif.h header requested but not found")
+      endif()
+  else()
+      # if f08 is not requested, then use f90 interface
+      if(MPI_Fortran_HAVE_F90_MODULE)
+          add_compile_definitions("MPI90")
+      else()
+          message(FATAL_ERROR "MPI f90 module not found")
+      endif()
+  endif()
+  add_compile_definitions("MPI")
+else()
+  message(STATUS "(serial build)")
+endif()
+
+find_package(BLAS REQUIRED) # binaries require blas
+if (BLAS_FOUND)
+    set(LIBS ${LIBS} ${BLAS_LIBRARIES})
+endif()
+
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
+# common objects
+set(src ${PROJECT_SOURCE_DIR}/src/comms.F90
+    ${PROJECT_SOURCE_DIR}/src/constants.F90
+    ${PROJECT_SOURCE_DIR}/src/error_base.F90
+    ${PROJECT_SOURCE_DIR}/src/error.F90
+    ${PROJECT_SOURCE_DIR}/src/io.F90
+    ${PROJECT_SOURCE_DIR}/src/kmesh.F90
+    ${PROJECT_SOURCE_DIR}/src/readwrite.F90
+    ${PROJECT_SOURCE_DIR}/src/types.F90
+    ${PROJECT_SOURCE_DIR}/src/utility.F90
+    ${PROJECT_SOURCE_DIR}/src/ws_distance.F90)
+
+# wannier90.x specific
+set(srcw90 ${PROJECT_SOURCE_DIR}/src/disentangle.F90
+    ${PROJECT_SOURCE_DIR}/src/hamiltonian.F90
+    ${PROJECT_SOURCE_DIR}/src/overlap.F90
+    ${PROJECT_SOURCE_DIR}/src/plot.F90
+    ${PROJECT_SOURCE_DIR}/src/sitesym.F90
+    ${PROJECT_SOURCE_DIR}/src/transport.F90
+    ${PROJECT_SOURCE_DIR}/src/wannier90_readwrite.F90
+    ${PROJECT_SOURCE_DIR}/src/wannier90_types.F90
+    ${PROJECT_SOURCE_DIR}/src/wannierise.F90)
+
+# postw90.x specific
+set(srcp90 ${PROJECT_SOURCE_DIR}/src/postw90/berry.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/boltzwann.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/dos.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/geninterp.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/get_oper.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/gyrotropic.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/kpath.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/kslice.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/postw90_common.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/postw90_readwrite.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/postw90_types.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/spin.F90
+    ${PROJECT_SOURCE_DIR}/src/postw90/wan_ham.F90)
+
+# library specific
+set(srclib ${PROJECT_SOURCE_DIR}/src/library_interface.F90)
+
+# library objects to avoid recompilation of common objects
+add_library(obj_common OBJECT ${src})
+add_library(obj_wann90 OBJECT ${srcw90} $<TARGET_OBJECTS:obj_common>)
+add_library(obj_post90 OBJECT ${srcp90} $<TARGET_OBJECTS:obj_common>)
+add_library(obj_lib OBJECT ${srclib} $<TARGET_OBJECTS:obj_common> $<TARGET_OBJECTS:obj_wann90> )
```

Comment:
Avoid using object libraries. Add the sources directly to the corresponding `add_executable`. In the case of `obj_common` it is better to link it directly to `wannier_lib`